### PR TITLE
[Doc] Workspace template examples

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This PR adds an initial set of examples that can be used as a template for users to get started with certain workloads.

This includes:
1. Batch prediction w/ AIR
2. Many model training w/ AIR
3. Serving a stable diffusion model

For each example, we include:
- Jupyter notebook
- (Optional) `requirements.txt` for additional dependencies
- Entry in `templates.yaml`
- Entry in release tests yaml

See [the contributing guide](https://github.com/ray-project/ray/blob/4940aa7830709920f93d93539bd7140eb5fa0033/doc/source/templates/README.md) to see the steps needed to add a new template.

## TODOs

- [x] Finish READMEs
- [x] Create hello world vs. production versions for all examples. Currently, there's just one version of each.
- [x] Add release tests

## To add in a follow-up PR

- [ ] Full Ray cluster launcher support
  - [ ] Guide on how to run these templates on OSS cluster
  - [ ] Show these templates somewhere in OSS docs
  - [ ] Create an script to convert OSS cluster config -> Anyscale configs

## Related issue number

N/A

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/examples/01_batch_inference/batch_inference.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    return {"image": transform(batch["image"]).numpy()}
+    # End
+
+
+preprocessor = BatchMapper(fn=preprocess, batch_format="numpy")
+
+# 3. Replace these with the framework of your choice!
+# Start
+predictor_cls = TorchPredictor
+checkpoint_cls = TorchCheckpoint
```
Comment (ericl) [inline doc/source/examples/01_batch_inference/batch_inference.py:56]: This is too framework specific. Can you change this to use more generic examples as in https://www.anyscale.com/blog/model-batch-inference-in-ray-actors-actorpool-and-datasets

To be clear, the issue is that the user has no idea how to write their own Predictor or Checkpoint given this template as an example--- all they get is "TorchPredictor"!

**Review by ericl — CHANGES_REQUESTED** [review]
Let's actually do two things for now:
1. Don't use BatchPredictor or Checkpoint APIs.
2. Instead, use the Dataset ActorPool APIs and callable classes.

We are having some internal discussions on possibly deprecating / unifying BatchPredictor: https://github.com/ray-project/ray/issues/32929

File: doc/source/examples/02_many_model_training/many_model_training.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# For reading from cloud storage paths.
+from smart_open import smart_open
+import pandas as pd
+
+import ray
+from ray import air, tune
+from ray.air import session
+
+ray.init()
```
Comment (ericl) [inline doc/source/examples/02_many_model_training/many_model_training.py:9]: Don't need ray.init.

File: doc/source/examples/02_many_model_training/many_model_training.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            "coef": reg.coef_,
+            "intercept": reg.intercept_,
+            "customer_id": data["customer_id"][0],
+            "score": score,
+        }
+    )
+
+
+# Tune is designed for up to thousands of trials.
+num_nodes = len(ray.nodes())
```
Comment (ericl) [inline doc/source/examples/02_many_model_training/many_model_training.py:34]: Remove num nodes.

File: doc/source/examples/02_many_model_training/many_model_training.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        ]
+    )
+}
+
+tuner = tune.Tuner(
+    trainable_func,
+    param_space=param_space,
+    run_config=air.RunConfig(
+        name="many_model_training_template",
+        local_dir="/mnt/cluster_storage/",
```
Comment (ericl) [inline doc/source/examples/02_many_model_training/many_model_training.py:51]: How about user or shared storage?

File: doc/source/examples/03_serving_stable_diffusion/app.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
+from io import BytesIO
+
+import ray
+from ray import serve
+from fastapi import FastAPI
+from fastapi.responses import Response
+
+
+NUM_NODES = len(ray.nodes())
```
Comment (ericl) [inline doc/source/examples/03_serving_stable_diffusion/app.py:9]: Don't query num nodes.

File: doc/source/examples/03_serving_stable_diffusion/app.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        assert len(prompt), "prompt parameter cannot be empty"
+
+        image = await (await self.handle.generate.remote(prompt, img_size=img_size))
+
+        file_stream = BytesIO()
+        image.save(file_stream, "PNG")
+        return Response(content=file_stream.getvalue(), media_type="image/png")
+
+
+def get_runtime_env():
```
Comment (ericl) [inline doc/source/examples/03_serving_stable_diffusion/app.py:35]: Don't use runtime envs.

File: doc/source/examples/03_serving_stable_diffusion/app.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    return {"pip": requirements}
+
+
+@serve.deployment(
+    ray_actor_options={"num_gpus": 1, "runtime_env": get_runtime_env()},
+    autoscaling_config={"min_replicas": NUM_NODES, "max_replicas": NUM_NODES},
+)
+class StableDiffusionV2:
+    def __init__(self):
+        # Delay imports, since packages will only be installed on the replicas
```
Comment (ericl) [inline doc/source/examples/03_serving_stable_diffusion/app.py:49]: Please move these imports to toplevel, and catch import / raise an error telling the user to install using `pip install --user` as in the README.

File: doc/source/examples/03_serving_stable_diffusion/app.py (hunk 6 above)
Comment (justinvyu) [reply doc/source/examples/03_serving_stable_diffusion/app.py:35]: I ended up using `runtime_env` due to `pip install --user` being really slow (due to EFS) to install the packages: 6 minutes compared to 1 minute for a regular `pip install`.

`runtime_env` is actually much faster to setup dependencies (it takes ~2 minutes to get the runtime env dependencies installed + get model pulled from huggingface).

Richard recommended this to get around introducing the runtime env concept:

> What if we don’t use the word runtime env, and instead used the wording of
“To start, we’ll set up our dependencies by passing this configuration to Ray. Ray will install all packages in this requirements list on each node.
There are other ways of installing dependencies that you can take a look at here (docs)”
ray.init(runtime_env={"pip": "./requirements.txt")

I'm okay with going with `pip install --user` if this initial buffer time of 5-8 minutes is not a dealbreaker. Personally, I've found it very nice to be able to run the script right away, without needing to install anything myself.

File: doc/source/examples/02_many_model_training/many_model_training.py (hunk 4 above)
Comment (justinvyu) [reply doc/source/examples/02_many_model_training/many_model_training.py:51]: Yep, can change it to `user_storage`.

If this mount doesn't exist (i.e. OSS cluster), I will default to `~/ray_results`.

File: doc/source/examples/02_many_model_training/many_model_training.py (hunk 3 above)
Comment (justinvyu) [reply doc/source/examples/02_many_model_training/many_model_training.py:34]: I'm using the number of nodes to distinguish between the hello world / larger scale examples. Is this too implicit - and you'd prefer providing a flag when running the script instead? Or is it becuase `ray.nodes()` is not something we should show the user?

File: doc/source/examples/03_serving_stable_diffusion/app.py (hunk 6 above)
Comment (ericl) [reply doc/source/examples/03_serving_stable_diffusion/app.py:35]: How about this instead?
1. We make the default config to use a single GPU node for the workspace.
2. Then, users can use normal `pip install` which is very fast. We can say use `pip install --user` or runtime_envs if you want to scale to a cluster.

Comment (justinvyu) [reply doc/source/examples/03_serving_stable_diffusion/app.py:35]: Sounds good, let's go with this!

File: doc/source/examples/01_batch_inference/batch_inference.ipynb [hunk 8]
```diff
@@ -0,0 +1,329 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "6fbc3e3c",
+   "metadata": {},
+   "source": [
+    "# Batch Inference with Ray Data\n",
```
Comment (ericl) [inline doc/source/examples/01_batch_inference/batch_inference.ipynb:8]: You're missing the install deps steps for the many model and batch inference examples.

**Review by ericl — COMMENTED** [review]
LGTM at a high level. The batch inference one will require some tuning as we enable streaming execution in 2.4, but we can cross that bridge as it comes.

File: doc/source/examples/01_batch_inference/batch_inference.ipynb (hunk 8 above)
Comment (justinvyu) [reply doc/source/examples/01_batch_inference/batch_inference.ipynb:8]: Added it for the MMT one. Batch inference doesn't have any not covered by ray-ml

File: doc/source/conf.py [hunk 9]
```diff
@@ -160,6 +160,7 @@
     "_build",
     "source/workflows/api/doc/ray.workflow.*",
     "source/serve/api/doc/ray.serve.*",
+    "source/templates",
```
Comment (richardliaw) [inline doc/source/conf.py:163]: maybe `source/templates/*`?

**Review by richardliaw — APPROVED** [review]

File: doc/source/templates/01_batch_inference/batch_inference.ipynb [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+    "tags": [
+     "small"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "NUM_WORKERS: int = 1\n",
+    "\n",
+    "USE_GPU: bool = True\n",
+    "NUM_GPUS_PER_WORKER: float = 1\n"
```
Comment (ericl) [inline doc/source/templates/01_batch_inference/batch_inference.ipynb:63]: This cell is repeated twice.

File: doc/source/templates/01_batch_inference/batch_inference.ipynb [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "23e77ada",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "preds = predictions.fully_executed()\n",
```
Comment (ericl) [inline doc/source/templates/01_batch_inference/batch_inference.ipynb:245]:
```suggestion
    "preds = predictions.cache()\n",
```

The fully executed call is deprecated, use cache() instead.

File: doc/source/templates/02_many_model_training/many_model_training.ipynb [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+   "execution_count": null,
+   "id": "a40e91a5",
+   "metadata": {
+    "tags": [
+     "small"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "NUM_DATA_PARTITIONS: int = 50\n"
```
Comment (ericl) [inline doc/source/templates/02_many_model_training/many_model_training.ipynb:142]: Duplicated cell. You can use a single cell with the other one commented out.

**Review by ericl — APPROVED** [review]

File: doc/source/templates/01_batch_inference/batch_inference.ipynb (hunk 10 above)
Comment (justinvyu) [reply doc/source/templates/01_batch_inference/batch_inference.ipynb:63]: This is how I'm getting around loading a small vs. large version of the template. On product side, we can run some post init command to process the notebook:
- Filter out all notebook cells tagged as `large`. Vice versa for large scale examples. This way, the configurations get set to use the small example defaults.

```bash
jupyter nbconvert --TagRemovePreprocessor.remove_input_tags='large'
      --to notebook --output batch_inference.ipynb batch_inference.ipynb
```

How does that sound @ericl?

Comment (justinvyu) [reply doc/source/templates/01_batch_inference/batch_inference.ipynb:63]: Yes, sounds good!

**Pull request merged**